text/0000-global-allocators.md

```diff
@@ -0,0 +1,248 @@
+(`#[proc_macro_derive]`)is used rather than a magic symbol name.
+
+Since an allocator is automatically selected when it is pulled into the crate
+graph, it is painful to compose allocators. For example, one may want to create
+an allocator which records statistics about active allocations, or adds padding
+around allocations to attempt to detect buffer overflows in unsafe code. To do
+this currently, the underlying allocator would need to be split into two
+crates, one which contains all of the functionality and another which is tagged
+as an `#![allocator]`.
+
+## jemalloc
+
+Rust's default allocator has historically been jemalloc. While jemalloc does
+provide significant speedups over certain system allocators for some allocation
+heavy workflows, it has has been a source of problems. For example, it has
+deadlock issues on Windows, does not work with Valgrind, adds ~300KB to
+binaries, and has caused crashes on macOS 10.12. See [this comment][] for more
+details. As a result, it is already disabled on many targets, including all of
+Windows. While there are certainly contexts in which jemalloc is a good choice,
+developers should be making that decision, not the compiler. The system
+allocator is a more reasonable and unsurprising default choice.
+
+A third party crate allowing users to opt-into jemalloc would also open the door
+to provide access to some of the library's other features such as tracing, arena
+pinning, and diagnostic output dumps for code that depends on jemalloc directly.
+
+[this comment]: https://github.com/rust-lang/rust/issues/36963#issuecomment-252029017
+
+# Detailed design
+[design]: #detailed-design
+
+## Defining an allocator
+
+An allocator crate identifies itself as such by applying the `#![allocator]`
+annotate at the crate root. It then defines a specific set of functions which
+are tagged with attributes:
+
+```rust
+#![allocator]
+
+/// Returns a pointer to `size` bytes of memory aligned to `align`.
+///
+/// On failure, returns a null pointer.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+#[allocator(allocate)]
+pub fn allocate(size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Returns a pointer to `size` bytes of memory aligned to `align`, and
+/// initialized with zeroes.
+///
+/// On failure, returns a null pointer.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+#[allocator(allocate_zeroed)]
+pub fn allocate_zeroed(size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Deallocates the memory referenced by `ptr`.
+///
+/// The `ptr` parameter must not be null.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
```

Comment:
That would be very relevant to both this RFC and the allocators design, so could you write up some details?